examples/helloworld-tutorial.md

```diff
@@ -0,0 +1,379 @@
+
+``` shell
+$ mkdir proto
+$ touch proto/helloworld.proto
+```
+
+Then the RPC methods get defined inside the service defintion, specifying the request and response types. gRPC lets you define four kinds of service methods, all of which are supported in OCaml. For this tutorial we will only use a simple RPC, if you would like to see an OCaml example which uses all four kinds please read the [routeguide tutorial].
+
+[routeguide tutorial]: https://github.com/dialohq/ocaml-grpc/blob/master/examples/routeguide-tutorial.md
+[Dune]: https://dune.readthedocs.io
+
+First we define our package name, which is the package OCaml looks for when including your protos in the client and server applications. Lets give this one a name of `helloworld`.
+
+``` protocol-buffer
+syntax = "proto3";
+package helloworld;
+```
+
+Next we need to define our service. This service will contain the actual RPC calls we will be using in our application. An RPC contains an Identifier, a Request type, and returns a Response type. Here is our Greeter service which provides the *SayHello* RPC method.
+
+``` protocol-buffer
+service Greeter {
+    // Our SayHello rpc accepts HelloRequests and returns HelloReplies
+    rpc SayHello (HelloRequest) returns (HelloReply);
+}
+```
+
+Next we have to define the types used in our `SayHello` RPC method. RPC types are defined as messages which contain typed fields. The messages for our HelloWorld application look like this.
+
+``` protocol-buffer
+message HelloRequest {
+    // Request message contains the name to be greeted
+    string name = 1;
+}
+
+message HelloReply {
+    // Reply contains the greeting message
+    string message = 1;
+}
+```
+
+Super! Now our `.proto` file should be complete and ready for use in our application. Here is what it should look like completed:
+
+``` protocol-buffer
+syntax = "proto3";
+package helloworld;
+
+service Greeter {
+    rpc SayHello (HelloRequest) returns (HelloReply);
+}
+
+message HelloRequest {
+   string name = 1;
+}
+
+message HelloReply {
+    string message = 1;
+}
+```
+
+## Application Setup 
+Now that we have defined the protobuf for our application we can start writing our OCaml. First we need to add our required dependencies to the `dune-project`.
+
+``` shell
+(lang dune 3.7)
+
+(name helloworld_grpc)
+(generate_opam_files true)
+
+(package
+ (name helloworld_grpc)
+ (synopsis "Helloworld in gRPC")
+ (description "Helloworld in gRPC.")
+ (depends
+  (grpc-lwt (>= 0.1.0))
+  (h2-lwt-unix (>= 0.9.0))
+  (ocaml-protoc (>= 2.4)))
+```
+
+We include `ocaml-protoc-plugin` as a useful way to incorporate the generation of our client and server gRPC code into the build process of our application. We will setup this build process now:
```

Comment:
Small nitpick: `ocaml-protoc` [1] and `ocaml-protoc-plugin` [2] are two distinct projects. Our gRPC library works with either (and also with other serialization formats like JSON) as long as you supply the encoder and decoder. The greeter examples uses `ocaml-protoc` while the etcd example uses `ocaml-protoc-plugin`.

[1] https://github.com/mransan/ocaml-protoc
[2] https://github.com/issuu/ocaml-protoc-plugin

Reply:
Thanks, that wasn't clear to me initially but it is now I've written some code with both. 
For example code I'd prefer suggesting people use one option that has broad support and point to a comparison with other options if they need it. The table on [issuu/ocaml-protoc-plugin](https://github.com/issuu/ocaml-protoc-plugin#comparison-with-other-ocaml-protobuf-handlers) has a comparison of the 3 protobuf options. In your opinion which one should be recommended for newcomers?

If you have an example of JSON serialisation, I would like to include that in the documentation as example code.

Reply:
> Thanks, that wasn't clear to me initially but it is now I've written some code with both. For example code I'd prefer suggesting people use one option that has broad support and point to a comparison with other options if they need it. The table on [issuu/ocaml-protoc-plugin](https://github.com/issuu/ocaml-protoc-plugin#comparison-with-other-ocaml-protobuf-handlers) has a comparison of the 3 protobuf options. In your opinion which one should be recommended for newcomers?

I'd say both `ocaml-protoc-plugin` and `ocaml-protoc` are fine. `ocaml-pb` seems less used and hasn't seen updates for a while, I think people are coalescing around the other two.

> If you have an example of JSON serialisation, I would like to include that in the documentation as example code.

I don't have any example code unfortunately. But the idea is that you can plug whichever format you like, since the serialization layer is completely independent of the transport layer. Here's a blog post about JSON specifically: https://grpc.io/blog/grpc-with-json/
That said, I haven't seen a JSON gRPC API in the wild, so I don't know if it would be helpful or confusing to mention this in a beginner's tutorial.